hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenGetNftInfoHandler.java

```diff
@@ -58,13 +71,75 @@ public Response createEmptyResponse(@NonNull final ResponseHeader header) {
     @Override
     public void validate(@NonNull final QueryContext context) throws PreCheckException {
         requireNonNull(context);
-        throw new UnsupportedOperationException("Not implemented");
+        final var query = context.query();
+        final var nftStore = context.createStore(ReadableNftStore.class);
+        final var op = query.tokenGetNftInfoOrThrow();
+        final var nftId = op.nftIDOrThrow();
+        validateTruePreCheck(nftId.hasTokenID(), INVALID_TOKEN_ID);
+        validateTruePreCheck(nftId.serialNumber() > 0, INVALID_TOKEN_NFT_SERIAL_NUMBER);
+
+        final var nft = nftStore.get(nftId.tokenID(), nftId.serialNumber());
+        validateFalsePreCheck(nft == null, INVALID_NFT_ID);
     }
 
     @Override
     public Response findResponse(@NonNull final QueryContext context, @NonNull final ResponseHeader header) {
         requireNonNull(context);
         requireNonNull(header);
-        throw new UnsupportedOperationException("Not implemented");
+        final var query = context.query();
+        final var config = context.configuration().getConfigData(LedgerConfig.class);
+        final var nftStore = context.createStore(ReadableNftStore.class);
+        final var op = query.tokenGetNftInfoOrThrow();
+        final var response = TokenGetNftInfoResponse.newBuilder();
+        final var nftId = op.nftIDOrElse(NftID.DEFAULT);
+
+        final var responseType = op.headerOrElse(QueryHeader.DEFAULT).responseType();
+        response.header(header);
+        if (header.nodeTransactionPrecheckCode() == OK && responseType != COST_ANSWER) {
+            final var optionalInfo = infoForNft(nftId, nftStore, config);
+            if (optionalInfo.isPresent()) {
+                response.nft(optionalInfo.get());
+            } else {
+                response.header(ResponseHeader.newBuilder()
+                        .nodeTransactionPrecheckCode(INVALID_NFT_ID)
+                        .cost(0)); // from mono service, need to validate in the future
```

Comment:
If cost is 0, why do we extend PaidQueryHandler instead of FreeQueryHandler?

Reply:
This code is in the if else section, which is we didn't find any TokenNftInfo. When this happened, mono service set this cost to 0.

Reply:
Thank you for explaining.